File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.*;

class Solution {
    public int solution(int[] priorities, int location) {
        int answer = 0;
        Queue<Process> processes = generateQueue(priorities);
        
        int mostPriority = getMostPriority(processes);
        while(!processes.isEmpty()) {
            Process process = processes.poll();
            if (process.isMostPriority(mostPriority)) {
                mostPriority = getMostPriority(processes);
                answer += 1;
                if (process.isTargetProcess(location)) {
                    break;
                }
            } else {
                processes.offer(process);
            }
            
        }
        
        return answer;
    }
    
    private Queue<Process> generateQueue(int[] priorities) {
        Queue<Process> processes = new LinkedList<>();
        for (int i = 0; i < priorities.length; i++) {
            processes.add(new Process(priorities[i], i));
        }
        
        return processes;
    }
    
    private int getMostPriority(Queue<Process> processes) {
        if (processes.size() == 0) {
            return 0;
        }
        return processes.stream()
            .map(process -> process.priority)
            .max(Integer::compareTo)
            .orElseThrow();
    }
    
    static class Process {
        int priority;
        int location;
        
        public Process(int priority, int location) {
            this.priority = priority;
            this.location = location;
        }
        
        public boolean isMostPriority(int mostPriority) {
            return this.priority == mostPriority;
        }
        
        public boolean isTargetProcess(int location) {
            return this.location == location;
        }
    }
}

File: 프로그래머스/2/42587. 프로세스/프로세스.py
from collections import deque

def solution(priorities, location):
    answer = 1
    sorted_p = sorted(priorities, reverse=True)
    print(sorted_p)
    
    work = []
    for i in range(len(priorities)):
        work.append([i, priorities[i]])
        
    q = deque(work)
    
    while q:
        work = q.popleft()
        if work[1] == sorted_p[0]:
            if work[0] == location:
                break
            sorted_p.pop(0)
            answer += 1
            continue
        q.append([work[0], work[1]])
    
    return answer